File: lite-app-starter/src/main/java/com/chimbori/liteapp/starter/InstallActivity.java
package com.chimbori.liteapp.starter;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Html;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK;
import static android.content.Intent.FLAG_ACTIVITY_MULTIPLE_TASK;
import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static android.content.Intent.FLAG_ACTIVITY_NO_HISTORY;

public class InstallActivity extends Activity implements View.OnClickListener {
  private static final String TAG = "InstallActivity";

  private Context context;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    context = getApplicationContext();
    setContentView(R.layout.activity_install_dialog);

    TextView descriptionTextView = findViewById(R.id.install_dialog_description);
    String descriptionHtml = getString(R.string.hermit_description);
    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
      descriptionTextView.setText(Html.fromHtml(descriptionHtml, Html.FROM_HTML_MODE_LEGACY));
    } else {
      descriptionTextView.setText(Html.fromHtml(descriptionHtml));
    }
    findViewById(R.id.install_dialog_install_button).setOnClickListener(this);
  }

  @Override
  public void onClick(View view) {
    if (view.getId() == R.id.install_dialog_install_button) {
      boolean wasPlayStoreSuccessful = navigateToPlayStore(this, Constants.HERMIT_PACKAGE_NAME);
      if (!wasPlayStoreSuccessful) {
        boolean wasBrowserNavigationSuccessful = navigateToHermitDownloadSite(this);
        if (!wasBrowserNavigationSuccessful) {
          Toast.makeText(context, context.getString(R.string.error_no_app_available), Toast.LENGTH_LONG).show();
        }
      }
      finishAndRemoveTask();
    } else if (view.getId() == R.id.install_dialog_cancel_button) {
      finishAndRemoveTask();
    }
  }

  private boolean navigateToHermitDownloadSite(@NonNull Activity activity) {
    Intent browserIntent = new Intent(Intent.ACTION_VIEW)
        .setData(Uri.parse(Constants.HERMIT_DOWNLOAD_LOCATION))
        .addFlags(FLAG_ACTIVITY_NO_HISTORY | FLAG_ACTIVITY_NEW_TASK | FLAG_ACTIVITY_CLEAR_TASK | FLAG_ACTIVITY_MULTIPLE_TASK);
    try {
      activity.startActivity(browserIntent);
      return true;

    } catch (ActivityNotFoundException e) {
      return false;
    }
  }

  private static boolean navigateToPlayStore(@NonNull Activity activity, @NonNull String packageId) {
    Context context = activity.getApplicationContext();
    Intent playStoreIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + packageId))
        .addFlags(FLAG_ACTIVITY_NO_HISTORY | FLAG_ACTIVITY_NEW_TASK | FLAG_ACTIVITY_CLEAR_TASK | FLAG_ACTIVITY_MULTIPLE_TASK);
    try {
      activity.startActivity(playStoreIntent);
      return true;
    } catch (ActivityNotFoundException anfe) {
      Toast.makeText(context, R.string.error_play_store, Toast.LENGTH_LONG).show();
      return false;
    }
  }
}

File: lite-app-starter/src/main/java/com/chimbori/liteapp/starter/InstallActivity.kt
package com.chimbori.liteapp.starter

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.Intent.ACTION_VIEW
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK
import android.content.Intent.FLAG_ACTIVITY_MULTIPLE_TASK
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.content.Intent.FLAG_ACTIVITY_NO_HISTORY
import android.net.Uri
import android.os.Build.VERSION.SDK_INT
import android.os.Build.VERSION_CODES.N
import android.os.Bundle
import android.text.Html
import android.text.Html.FROM_HTML_MODE_LEGACY
import android.view.View
import android.widget.TextView
import android.widget.Toast
import android.widget.Toast.LENGTH_LONG
import com.chimbori.liteapp.starter.Constants.HERMIT_DOWNLOAD_LOCATION
import com.chimbori.liteapp.starter.Constants.HERMIT_PACKAGE_NAME

class InstallActivity : Activity(), View.OnClickListener {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_install_dialog)

    findViewById<TextView>(R.id.install_dialog_description).text = if (SDK_INT >= N) {
      Html.fromHtml(getString(R.string.hermit_description), FROM_HTML_MODE_LEGACY)
    } else {
      Html.fromHtml(getString(R.string.hermit_description))
    }
    findViewById<View>(R.id.install_dialog_install_button).setOnClickListener(this)
    findViewById<View>(R.id.install_dialog_cancel_button).setOnClickListener(this)
  }

  override fun onClick(view: View) {
    when (view.id) {
      R.id.install_dialog_install_button -> {
        val wasPlayStoreSuccessful = navigateToPlayStore(this, HERMIT_PACKAGE_NAME)
        if (!wasPlayStoreSuccessful) {
          val wasBrowserNavigationSuccessful = navigateToHermitDownloadSite(this)
          if (!wasBrowserNavigationSuccessful) {
            Toast.makeText(applicationContext, getString(R.string.error_no_app_available), LENGTH_LONG).show()
          }
        }
        finishAndRemoveTask()
      }
      R.id.install_dialog_cancel_button -> finishAndRemoveTask()
    }
  }

  companion object {
    private fun navigateToHermitDownloadSite(activity: Activity): Boolean = try {
      activity.startActivity(
          Intent(ACTION_VIEW).setData(Uri.parse(HERMIT_DOWNLOAD_LOCATION))
              .addFlags(FLAG_ACTIVITY_NO_HISTORY or FLAG_ACTIVITY_NEW_TASK or FLAG_ACTIVITY_CLEAR_TASK or FLAG_ACTIVITY_MULTIPLE_TASK))
      true
    } catch (e: ActivityNotFoundException) {
      false
    }

    private fun navigateToPlayStore(activity: Activity, packageId: String): Boolean {
      return try {
        activity.startActivity(Intent(ACTION_VIEW, Uri.parse("market://details?id=$packageId"))
            .addFlags(FLAG_ACTIVITY_NO_HISTORY or FLAG_ACTIVITY_NEW_TASK or FLAG_ACTIVITY_CLEAR_TASK or FLAG_ACTIVITY_MULTIPLE_TASK))
        true
      } catch (e: ActivityNotFoundException) {
        Toast.makeText(activity.applicationContext, R.string.error_play_store, LENGTH_LONG).show()
        false
      }
    }
  }
}